numhask/src/NumHask/Algebra/Group.hs

```diff
@@ -0,0 +1,117 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ConstrainedClassMethods #-}
+{-# LANGUAGE RoleAnnotations #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE MonoLocalBinds #-}
+-- | The Group hirarchy
+module Numhask.Algebra.Group
+      ( Magma(..)
+      , Unital(..)
+      , Semigroup(..)
+      , Commutative
+      , Absorbing(..)
+      , Invertible(..)
+      , Idempotent
+      , Monoid(..)
+      , Group
+      , groupSwap
+      , AbelianGroup
+      )
+      where
+
+import qualified Prelude                       as P
+
+-- * Magma structure
+-- | A <https://en.wikipedia.org/wiki/Magma_(algebra) Magma> is a tuple (T,comb) consisting of
+--
+-- - a type a, and
+--
+-- - a function (comb) :: T -> T -> T
+--
+-- The mathematical laws for a magma are:
+--
+-- - comb is defined for all possible pairs of type T, and
+--
+-- - comb is closed in the set of all possible values of type T
+--
+-- or, more tersly,
+--
+-- > ∀ a, b ∈ T: a comb b ∈ T
+--
+-- These laws are true by construction in haskell: the type signature of 'magma' and the above mathematical laws are synonyms.
+--
+--
+class Magma a where
+  comb :: a -> a -> a
+
+-- | A Unital Magma
+--
+-- > unit comb a = a
+-- > a comb unit = a
+--
+class Magma a =>
+      Unital a where
+  unit :: a
+
+-- | A semigroup is an associative Magma
+--
+-- > (a comb b) comb c = a comb (b comb c)
+class Magma a =>
+      Semigroup a where
```

Comment:
Why do we need to redefine Semigroup and Monoid ?

Reply:
that's a good question actually. I am not sure whether we need to redefine it.
I've redefined it because we get the nice Magma-superclass, but I think a more pragmatic approach may be better (just use it and don't bother about magma).